src/auth/authenticator/tls/authenticator.go

```diff
@@ -0,0 +1,51 @@
+package tls
+
+import (
+	"crypto/x509"
+	"net/http"
+
+	"github.com/consensys/quorum-key-manager/pkg/errors"
+	"github.com/consensys/quorum-key-manager/src/auth/types"
+)
+
+const AuthMode = "Tls"
+
+type Authenticator struct {
+	Certificates []*x509.Certificate
+}
+
+func NewAuthenticator(cfg *Config) (*Authenticator, error) {
+	if len(cfg.Certificates) == 0 {
+		return nil, nil
+	}
+
+	auth := &Authenticator{Certificates: cfg.Certificates}
+
+	return auth, nil
+}
+
+// Authenticate checks certs and retrieve user Info
+// CN field -> Username
+// Organization -> Groups
+func (authenticator Authenticator) Authenticate(req *http.Request) (*types.UserInfo, error) {
+	// extract Certificate info from request if any
+	if len(req.TLS.PeerCertificates) == 0 {
+		return nil, nil
+	}
+	// first array element is the leaf
+	clientCert := *req.TLS.PeerCertificates[0]
+	// check this cert matches authenticator provided one
+	// using strict comparison
+	for _, authCert := range authenticator.Certificates {
+		if clientCert.Equal(authCert) {
+			return &types.UserInfo{
+				Username: clientCert.Subject.CommonName,
+				Groups:   clientCert.Subject.Organization,
+				AuthMode: AuthMode,
+			}, nil
+
+		}
+	}
+
+	return nil, errors.UnauthorizedError("no matching cert found")
```

Comment:
As far as I see we should diferenciate between using a standard tls connection to protect client to server communications, which will be validated by your http server, and using a tls certificate to authenticate your connection including username and groups. That in why I believe it is important to dive into the provided chain of certs in this middleware.

 In addition I think we should not reject in case of invalid certificate because of those reasons. wdyt @nmvalera

Reply:
As of today certificate matching is no longer in place.
Every valid and provided cert will be treated as a correct authentication